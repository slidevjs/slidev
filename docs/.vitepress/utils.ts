import { data as features } from '../features/index.data.js'
import { Advanced, Guides } from './pages'

function removeHash(link: string) {
  const idx = link.lastIndexOf('#')
  return idx < 0 ? link : link.slice(0, idx)
}

function getGuideTitle(id: string) {
  return Guides.find(g => g.link.endsWith(`/${id}`))?.text ?? Advanced.find(g => g.link.endsWith(id))?.text ?? id
}

export function resolveLink(link: string): {
  kind: 'external' | 'features' | 'guide'
  url: string
  title?: string
  tags?: string[]
  descripton?: string
} {
  const [kind, nameWithHash] = link.split('/')
  const name = removeHash(nameWithHash)
  switch (kind) {
    case 'http:':
    case 'https:':
    case 'mailto:':
      return { kind: 'external', url: link }
    case 'features': {
      const feature = features[name]
      if (!feature)
        throw new Error(`Feature "${name}" not found.`)
      return {
        kind: 'features',
        title: `âœ¨ ${feature.title}`,
        tags: feature.tags,
        descripton: feature.description,
        url: `/features/${nameWithHash}`,
      }
    }
    case 'guide': {
      return {
        kind: 'guide',
        title: `ðŸ“–  ${getGuideTitle(name)}`,
        tags: ['guide'],
        descripton: 'Click to read this guide',
        url: `/guide/${nameWithHash}`,
      }
    }
    default:
      throw new Error(`Invalid link: ${link}`)
  }
}
