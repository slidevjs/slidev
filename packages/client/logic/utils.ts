import { useTimestamp } from '@vueuse/core'
import { computed, ref } from 'vue'

export function useTimer() {
  const tsStart = ref(Date.now())
  const now = useTimestamp({
    interval: 1000,
  })
  const timer = computed(() => {
    const passed = (now.value - tsStart.value) / 1000
    const sec = Math.floor(passed % 60).toString().padStart(2, '0')
    const min = Math.floor(passed / 60).toString().padStart(2, '0')
    return `${min}:${sec}`
  })
  function resetTimer() {
    tsStart.value = now.value
  }

  return {
    timer,
    resetTimer,
  }
}

export function makeId(length = 5) {
  const result = []
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  const charactersLength = characters.length
  for (let i = 0; i < length; i++)
    result.push(characters.charAt(Math.floor(Math.random() * charactersLength)))
  return result.join('')
}
