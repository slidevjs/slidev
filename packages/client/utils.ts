import type { RouteRecordRaw } from 'vue-router'

export function getSlideClass(route?: RouteRecordRaw, extra = '') {
  const classes = ['slidev-page', extra]

  const no = route?.meta?.slide?.no
  if (no != null)
    classes.push(`slidev-page-${no}`)

  return classes.filter(Boolean).join(' ')
}
