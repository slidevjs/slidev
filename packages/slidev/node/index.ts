import './declare'

export * from './server'
export * from './plugins/preset'
export * from './options'
export { parser } from './parser'
