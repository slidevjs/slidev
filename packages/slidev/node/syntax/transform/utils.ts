export function normalizeRangeStr(rangeStr = '') {
  return !rangeStr.trim() ? [] : rangeStr.trim().split(/\|/g).map(i => i.trim())
}

/**
 * Escape `{{` in code block to prevent Vue interpret it, #99, #1316
 */
export function escapeVueInCode(md: string) {
  return md.replace(/{{/g, '&lbrace;&lbrace;')
}
