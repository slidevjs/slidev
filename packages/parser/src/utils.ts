import { isNumber, range, uniq } from '@antfu/utils'

/**
 * 1,3-5,8 => [1, 3, 4, 5, 8]
 */
export function parseRangeString(total: number, rangeStr?: string) {
  if (!rangeStr || rangeStr === 'all' || rangeStr === '*')
    return range(1, total + 1)

  const pages: number[] = []
  for (const part of rangeStr.split(/[,;]/g)) {
    if (!part.includes('-')) {
      pages.push(+part)
    }
    else {
      const [start, end] = part.split('-', 2)
      pages.push(
        ...range(+start, !end ? (total + 1) : (+end + 1)),
      )
    }
  }

  return uniq(pages).filter(i => i <= total).sort((a, b) => a - b)
}

/**
 * Accepts `16/9` `1:1` `3x4`
 */
export function parseAspectRatio(str: string | number) {
  if (isNumber(str))
    return str
  if (!Number.isNaN(+str))
    return +str
  const [wStr = '', hStr = ''] = str.split(/[:\/x\|]/)
  const w = Number.parseFloat(wStr.trim())
  const h = Number.parseFloat(hStr.trim())

  if (Number.isNaN(w) || Number.isNaN(h) || h === 0)
    throw new Error(`Invalid aspect ratio "${str}"`)

  return w / h
}
