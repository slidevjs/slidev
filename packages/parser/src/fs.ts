import { promises as fs } from 'node:fs'
import { dirname, resolve } from 'node:path'
import YAML from 'js-yaml'
import { slash } from '@antfu/utils'
import type { PreparserExtensionLoader, SlideInfo, SlidevData, SlidevMarkdown, SlidevPreparserExtension, SourceSlideInfo } from '@slidev/types'
import { detectFeatures, parse, parseRangeString, stringify } from './core'

export * from './core'

let preparserExtensionLoader: PreparserExtensionLoader | null = null

export function injectPreparserExtensionLoader(fn: PreparserExtensionLoader) {
  preparserExtensionLoader = fn
}

/**
 * Slidev data without config and themeMeta,
 * because config and themeMeta depends on the theme to be loaded.
 */
export type LoadedSlidevData = Omit<SlidevData, 'config' | 'themeMeta'>

export async function load(userRoot: string, filepath: string, content?: string, mode?: string): Promise<LoadedSlidevData> {
  const markdown = content ?? await fs.readFile(filepath, 'utf-8')

  let extensions: SlidevPreparserExtension[] | undefined
  if (preparserExtensionLoader) {
    // #703
    // identify the headmatter, to be able to load preparser extensions
    // (strict parsing based on the parsing code)
    const lines = markdown.split(/\r?\n/g)
    let hm = ''
    if (lines[0].match(/^---([^-].*)?$/) && !lines[1]?.match(/^\s*$/)) {
      let hEnd = 1
      while (hEnd < lines.length && !lines[hEnd].trimEnd().match(/^---$/))
        hEnd++
      hm = lines.slice(1, hEnd).join('\n')
    }
    const o = YAML.load(hm) as Record<string, unknown> ?? {}
    extensions = await preparserExtensionLoader(o, filepath, mode)
  }

  const markdownFiles: Record<string, SlidevMarkdown> = {}
  const slides: SlideInfo[] = []

  async function loadMarkdown(path: string, range?: string, frontmatterOverride?: Record<string, unknown>) {
    let md = markdownFiles[path]
    if (!md) {
      const raw = await fs.readFile(path, 'utf-8')
      md = await parse(raw, path, extensions)
      markdownFiles[path] = md
    }

    for (const index of parseRangeString(md.slides.length, range))
      await loadSlide(md.slides[index - 1], frontmatterOverride)

    return md
  }

  async function loadSlide(slide: SourceSlideInfo, frontmatterOverride?: Record<string, unknown>) {
    if (slide.frontmatter.disabled || slide.frontmatter.hide)
      return
    if (slide.frontmatter.src) {
      const [rawPath, rangeRaw] = slide.frontmatter.src.split('#')
      const path = rawPath.startsWith('/')
        ? resolve(userRoot, rawPath.substring(1))
        : resolve(dirname(slide.filepath), rawPath)

      frontmatterOverride = {
        ...slide.frontmatter,
        ...frontmatterOverride,
      }
      delete frontmatterOverride.src

      await loadMarkdown(path, rangeRaw, frontmatterOverride)
    }
    else {
      slides.push({
        index: slides.length,
        source: slide,
        frontmatter: { ...slide.frontmatter, ...frontmatterOverride },
        content: slide.content,
        note: slide.note,
        title: slide.title,
        level: slide.level,
      })
    }
  }

  const entry = await loadMarkdown(filepath)

  const headmatter = {
    title: slides[0]?.title,
    ...entry.slides[0]?.frontmatter,
  }

  return {
    slides,
    entry,
    headmatter,
    features: detectFeatures(slides.map(s => s.source.raw).join('')),
    markdownFiles,
    watchFiles: Object.keys(markdownFiles).map(slash),
  }
}

export async function save(markdown: SlidevMarkdown) {
  await fs.writeFile(markdown.filepath, stringify(markdown), 'utf-8')
}
