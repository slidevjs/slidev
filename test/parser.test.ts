import { basename, relative, resolve } from 'node:path'
import fg from 'fast-glob'
import { describe, expect, it } from 'vitest'
import { objectMap, slash } from '@antfu/utils'
import { getDefaultConfig, load, parse, prettify, resolveConfig, stringify } from '../packages/parser/src/fs'
import type { SlidevConfig, SlidevPreparserExtension } from '../packages/types/src'

function configDiff(v: SlidevConfig) {
  const defaults = getDefaultConfig()
  const res: Record<string, any> = {}
  for (const key of Object.keys(v) as (keyof SlidevConfig)[]) {
    if (JSON.stringify(v[key]) !== JSON.stringify(defaults[key]))
      res[key] = v[key]
  }
  return res
}

function replaceCRLF(s: string) {
  return s.replace(/\r\n/g, '\n')
}

describe('md parser', () => {
  const userRoot = resolve(__dirname, 'fixtures/markdown')
  const files = fg.sync('*.md', {
    cwd: userRoot,
    absolute: true,
  })

  for (const file of files) {
    it(basename(file), async () => {
      const data = await load(userRoot, file)

      expect(stringify(data.entry).trim()).toEqual(replaceCRLF(data.entry.raw.trim()))

      prettify(data.entry)

      // File path tests & convert to relative paths
      data.markdownFiles = objectMap(data.markdownFiles, (path, md) => {
        expect(md.filepath).toBe(path)
        const relativePath = slash(relative(userRoot, path))
        md.slides.forEach((slide) => {
          expect(slide.filepath).toBe(path)
          slide.filepath = relativePath
        })
        md.filepath = relativePath
        return [relativePath, md]
      })

      expect(data.slides).toMatchSnapshot('slides')
      expect(configDiff(resolveConfig(data.headmatter, {}))).toMatchSnapshot('config')
      expect(data.features).toMatchSnapshot('features')
    })
  }

  it('parse', async () => {
    const data = await parse(`
a

---

b

---
layout: z
---
c
----
d
----
e

---

f

`, 'file.md')
    expect(data.slides.map(i => i.content.trim()))
      .toEqual(Array.from('abcdef'))
    expect(data.slides[2].frontmatter)
      .toEqual({ layout: 'z' })
    expect(data.slides[3].frontmatter)
      .toEqual({ })
  })

  it('parse section matter', async () => {
    const data = await parse(`
a

---

b

---section2
layout: z
---
c
----   section 3
d
---- section-4
e

---

f

`, 'file.md')
    expect(data.slides.map(i => i.content.trim()))
      .toEqual(Array.from('abcdef'))
    expect(data.slides[2].frontmatter)
      .toEqual({ layout: 'z' })
    expect(data.slides[3].frontmatter)
      .toEqual({ })
  })

  async function parseWithExtension(
    src: string,
    transformRawLines: (lines: string[]) => void | Promise<void> = () => {},
    more = {},
    moreExts: SlidevPreparserExtension[] = [],
  ) {
    return await parse(
      src,
      'file.md',
      [{ transformRawLines, ...more }, ...moreExts] as any,
    )
  }

  it('parse with-extension replace', async () => {
    const data = await parseWithExtension(`---
ga: bu
---
a @@v@@

---

b
@@v@@

@@v@@ = @@v@@
`, (lines) => {
      for (const i in lines)
        lines[i] = lines[i].replace(/@@v@@/g, 'thing')
    })

    expect(data.slides.map(i => i.content.trim().replace(/\n/g, '%')).join('/'))
      .toEqual('a thing/b%thing%%thing = thing')
  })

  it('parse with-extension custom-separator', async () => {
    const data = await parseWithExtension(`---
ga: bu
---
a @@v@@

SEPARATOR

b
@@v@@

@@v@@ = @@v@@
`, (lines) => {
      for (const i in lines)
        lines[i] = lines[i].replace(/^SEPARATOR$/g, '---')
    })

    expect(data.slides.map(i => i.content.trim().replace(/\n/g, '%')).join('/'))
      .toEqual('a @@v@@/b%@@v@@%%@@v@@ = @@v@@')
  })

  it('parse with-extension eg-easy-cover', async () => {
    function cov(i: string, more = '.jpg') {
      return `---
layout: cover
background: ${i}${more}
---

`
    }
    const data = await parseWithExtension(`@cov 1.jpg
@cov 2.jpg
# 2
---

# 3
@cov 4.jpg
`, (lines) => {
      let i = 0
      while (i < lines.length) {
        if (lines[i].startsWith('@cov ')) {
          const repl = [...cov(lines[i].substring(5), '').split('\n')]
          lines.splice(i, 1, ...repl)
        }
        i++
      }
    })

    expect(data.slides.map(s => s.content.trim().replace(/\n/g, '%')).join('/'))
      .toEqual('/# 2/# 3/'.replace(/\n/g, '%'))
    expect(data.slides[0].frontmatter)
      .toEqual({ layout: 'cover', background: '1.jpg' })
    expect(data.slides[1].frontmatter)
      .toEqual({ layout: 'cover', background: '2.jpg' })
    expect(data.slides[3].frontmatter)
      .toEqual({ layout: 'cover', background: '4.jpg' })
  })

  it('parse with-extension sequence', async () => {
    const data = await parseWithExtension(`
a..A
a.a.A.A
.a.A.
`, undefined, {}, [{
      name: 'test',
      transformRawLines(lines: string[]) {
        for (const i in lines)
          lines[i] = lines[i].replace(/A/g, 'B').replace(/a/g, 'A')
      },
    }, {
      name: 'test',
      transformRawLines(lines: string[]) {
        for (const i in lines)
          lines[i] = lines[i].replace(/A/g, 'C')
      },
    }])
    expect(data.slides.map(i => i.content.trim().replace(/\n/g, '%')).join('/'))
      .toEqual('C..B%C.C.B.B%.C.B.')
  })

  // Generate cartesian product of given iterables:
  function* cartesian(...all: any[]): Generator<any[], void, void> {
    const [head, ...tail] = all
    const remainder = tail.length ? cartesian(...tail) : [[]]
    for (const r of remainder) {
      for (const h of head)
        yield [h, ...r]
    }
  }
  const B = [0, 1]
  const Bs = [B, B, B, B, B, B]
  const bNames = '_swScCF'

  for (const desc of cartesian(...Bs)) {
    const [withSlideBefore, withFrontmatter, withSlideAfter, prependContent, appendContent, addFrontmatter] = desc
    it(`parse with-extension wrap ${desc.map((b, i) => bNames[b * (i + 1)]).join('')}`, async () => {
      const data = await parseWithExtension(`${
withSlideBefore
? `
.

---
`
: ''}${(!withSlideBefore && withFrontmatter) ? '---\n' : ''}${withFrontmatter
? `m: M
n: N
---`
: ''}

ccc
@a
@b
ddd
${
withSlideAfter
? `
---

..
`
: ''}`, undefined, {
        transformSlide(content: string, frontmatter: any) {
          const lines = content.split('\n')
          let i = 0
          let appendBeforeCount = 0
          let appendAfterCount = 0
          while (i < lines.length) {
            const l = lines[i]
            if (l.startsWith('@')) {
              const t = l.substring(1)
              lines.splice(i, 1)
              if (prependContent)
                lines.splice(appendBeforeCount++, 0, `<${t}>`)
              if (appendContent)
                lines.splice(lines.length - appendAfterCount++, 0, `</${t}>`)
              if (addFrontmatter)
                frontmatter[`add${t}`] = 'add'
              i--
            }
            i++
          }
          return lines.join('\n')
        },
      })
      function project(s: string) {
        // like the trim in other tests, the goal is not to test newlines here
        return s.replace(/%%%*/g, '%')
      }
      expect(project(data.slides.map(i => i.content.replace(/\n/g, '%')).join('/')))
        .toEqual(project([
          ...withSlideBefore ? ['./'] : [],
          ...prependContent ? ['<a>%<b>%'] : [],
          'ccc%ddd',
          ...appendContent ? ['%</b>%</a>'] : [],
          ...withSlideAfter ? ['/..'] : [],
        ].join('')))

      if (withFrontmatter || addFrontmatter) {
        expect(data.slides[withSlideBefore ? 1 : 0].frontmatter)
          .toEqual({
            ...withFrontmatter ? { m: 'M', n: 'N' } : {},
            ...addFrontmatter ? { adda: 'add', addb: 'add' } : {},
          })
      }
    })
  }
})
