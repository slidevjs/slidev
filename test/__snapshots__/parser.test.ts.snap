// Vitest Snapshot v1

exports[`md parser > frontmatter.md > config 1`] = `
{
  "addons": [],
  "aspectRatio": 1.7777777777777777,
  "canvasWidth": 980,
  "codeCopy": true,
  "colorSchema": "auto",
  "css": "windicss",
  "download": false,
  "drawings": {
    "enabled": true,
    "persist": false,
    "presenterOnly": false,
    "syncAll": true,
  },
  "exportFilename": "",
  "favicon": "https://cdn.jsdelivr.net/gh/slidevjs/slidev/assets/favicon.png",
  "fonts": {
    "italic": false,
    "local": [],
    "mono": [
      "\\"Fira Code\\"",
      "ui-monospace",
      "SFMono-Regular",
      "Menlo",
      "Monaco",
      "Consolas",
      "\\"Liberation Mono\\"",
      "\\"Courier New\\"",
      "monospace",
    ],
    "provider": "google",
    "sans": [
      "\\"Roboto\\"",
      "\\"Lato\\"",
      "ui-sans-serif",
      "system-ui",
      "-apple-system",
      "BlinkMacSystemFont",
      "\\"Segoe UI\\"",
      "Roboto",
      "\\"Helvetica Neue\\"",
      "Arial",
      "\\"Noto Sans\\"",
      "sans-serif",
      "\\"Apple Color Emoji\\"",
      "\\"Segoe UI Emoji\\"",
      "\\"Segoe UI Symbol\\"",
      "\\"Noto Color Emoji\\"",
    ],
    "serif": [
      "\\"Mate SC\\"",
      "ui-serif",
      "Georgia",
      "Cambria",
      "\\"Times New Roman\\"",
      "Times",
      "serif",
    ],
    "webfonts": [
      "Roboto",
      "Lato",
      "Mate SC",
      "Fira Code",
    ],
    "weights": [
      "200",
      "400",
      "600",
    ],
  },
  "highlighter": "prism",
  "info": false,
  "layout": "cover",
  "lineNumbers": false,
  "monaco": "dev",
  "plantUmlServer": "https://www.plantuml.com/plantuml",
  "record": "dev",
  "remoteAssets": false,
  "routerMode": "history",
  "selectable": false,
  "theme": "default",
  "themeConfig": {},
  "title": "Hi",
  "titleTemplate": "%s - Slidev",
}
`;

exports[`md parser > frontmatter.md > features 1`] = `
{
  "katex": false,
  "mermaid": false,
  "monaco": false,
  "tweet": false,
}
`;

exports[`md parser > frontmatter.md > slides 1`] = `
[
  {
    "content": "
# Hi
",
    "end": 8,
    "frontmatter": {
      "fonts": {
        "mono": "Fira Code",
        "sans": "Roboto, Lato",
        "serif": "Mate SC",
      },
      "layout": "cover",
      "title": "Hi",
    },
    "index": 0,
    "level": 1,
    "note": undefined,
    "raw": "---
layout: cover
fonts:
  sans: Roboto, Lato
  serif: Mate SC
  mono: Fira Code
title: Hi
---

# Hi

",
    "start": 0,
    "title": "Hi",
  },
  {
    "content": "
# Hello
",
    "end": 16,
    "frontmatter": {
      "layout": "center",
      "meta": {
        "duration": 12,
        "title": "FooBar",
      },
    },
    "index": 1,
    "level": 1,
    "note": "This is note",
    "raw": "---
meta:
  title: FooBar
  duration: 12
layout: center
---

# Hello

<!--
This is note
-->
",
    "start": 8,
    "title": "Hello",
  },
  {
    "content": "
# Morning
",
    "end": 19,
    "frontmatter": {},
    "index": 2,
    "level": 1,
    "note": undefined,
    "raw": "
# Morning

",
    "start": 17,
    "title": "Morning",
  },
  {
    "content": "
<!-- This is not note -->
Hey
",
    "end": 26,
    "frontmatter": {
      "layout": "text",
    },
    "index": 3,
    "level": undefined,
    "note": "This is note",
    "raw": "---
layout: text
---

<!-- This is not note -->
Hey

<!--
This is note
-->
",
    "start": 19,
    "title": undefined,
  },
  {
    "content": "
\`\`\`md
---
this should be treated as code block
---
----
\`\`\`
",
    "end": 35,
    "frontmatter": {},
    "index": 4,
    "level": undefined,
    "note": undefined,
    "raw": "
\`\`\`md
---
this should be treated as code block
---
----
\`\`\`

",
    "start": 27,
    "title": undefined,
  },
]
`;

exports[`md parser > minimal.md > config 1`] = `
{
  "addons": [],
  "aspectRatio": 1.7777777777777777,
  "canvasWidth": 980,
  "codeCopy": true,
  "colorSchema": "auto",
  "css": "windicss",
  "download": false,
  "drawings": {
    "enabled": true,
    "persist": false,
    "presenterOnly": false,
    "syncAll": true,
  },
  "exportFilename": "",
  "favicon": "https://cdn.jsdelivr.net/gh/slidevjs/slidev/assets/favicon.png",
  "fonts": {
    "italic": false,
    "local": [],
    "mono": [
      "ui-monospace",
      "SFMono-Regular",
      "Menlo",
      "Monaco",
      "Consolas",
      "\\"Liberation Mono\\"",
      "\\"Courier New\\"",
      "monospace",
    ],
    "provider": "google",
    "sans": [
      "ui-sans-serif",
      "system-ui",
      "-apple-system",
      "BlinkMacSystemFont",
      "\\"Segoe UI\\"",
      "Roboto",
      "\\"Helvetica Neue\\"",
      "Arial",
      "\\"Noto Sans\\"",
      "sans-serif",
      "\\"Apple Color Emoji\\"",
      "\\"Segoe UI Emoji\\"",
      "\\"Segoe UI Symbol\\"",
      "\\"Noto Color Emoji\\"",
    ],
    "serif": [
      "ui-serif",
      "Georgia",
      "Cambria",
      "\\"Times New Roman\\"",
      "Times",
      "serif",
    ],
    "webfonts": [],
    "weights": [
      "200",
      "400",
      "600",
    ],
  },
  "highlighter": "prism",
  "info": false,
  "lineNumbers": false,
  "monaco": "dev",
  "plantUmlServer": "https://www.plantuml.com/plantuml",
  "record": "dev",
  "remoteAssets": false,
  "routerMode": "history",
  "selectable": false,
  "theme": "default",
  "themeConfig": {},
  "title": "H1",
  "titleTemplate": "%s - Slidev",
}
`;

exports[`md parser > minimal.md > features 1`] = `
{
  "katex": false,
  "mermaid": false,
  "monaco": false,
  "tweet": false,
}
`;

exports[`md parser > minimal.md > slides 1`] = `
[
  {
    "content": "
# H1
## H2
### H3

Sample Text

\`\`\`ts
console.log('Hello World')
\`\`\`
",
    "end": 10,
    "frontmatter": {
      "title": "H1",
    },
    "index": 0,
    "level": 1,
    "note": undefined,
    "raw": "---
title: H1
---

# H1
## H2
### H3

Sample Text

\`\`\`ts
console.log('Hello World')
\`\`\`

",
    "start": 0,
    "title": "H1",
  },
  {
    "content": "
# Hello

- Hello
- Hi
- Hey
- Yo
",
    "end": 18,
    "frontmatter": {},
    "index": 1,
    "level": 1,
    "note": undefined,
    "raw": "
# Hello

- Hello
- Hi
- Hey
- Yo

",
    "start": 11,
    "title": "Hello",
  },
  {
    "content": "
Nice to meet you
",
    "end": 22,
    "frontmatter": {},
    "index": 2,
    "level": undefined,
    "note": undefined,
    "raw": "
Nice to meet you

",
    "start": 19,
    "title": undefined,
  },
]
`;

exports[`md parser > multi-entries.md > config 1`] = `
{
  "addons": [],
  "aspectRatio": 1.7777777777777777,
  "canvasWidth": 980,
  "codeCopy": true,
  "colorSchema": "auto",
  "css": "windicss",
  "download": false,
  "drawings": {
    "enabled": true,
    "persist": false,
    "presenterOnly": false,
    "syncAll": true,
  },
  "exportFilename": "",
  "favicon": "https://cdn.jsdelivr.net/gh/slidevjs/slidev/assets/favicon.png",
  "fonts": {
    "italic": false,
    "local": [],
    "mono": [
      "ui-monospace",
      "SFMono-Regular",
      "Menlo",
      "Monaco",
      "Consolas",
      "\\"Liberation Mono\\"",
      "\\"Courier New\\"",
      "monospace",
    ],
    "provider": "google",
    "sans": [
      "ui-sans-serif",
      "system-ui",
      "-apple-system",
      "BlinkMacSystemFont",
      "\\"Segoe UI\\"",
      "Roboto",
      "\\"Helvetica Neue\\"",
      "Arial",
      "\\"Noto Sans\\"",
      "sans-serif",
      "\\"Apple Color Emoji\\"",
      "\\"Segoe UI Emoji\\"",
      "\\"Segoe UI Symbol\\"",
      "\\"Noto Color Emoji\\"",
    ],
    "serif": [
      "ui-serif",
      "Georgia",
      "Cambria",
      "\\"Times New Roman\\"",
      "Times",
      "serif",
    ],
    "webfonts": [],
    "weights": [
      "200",
      "400",
      "600",
    ],
  },
  "highlighter": "prism",
  "info": false,
  "lineNumbers": false,
  "monaco": "dev",
  "plantUmlServer": "https://www.plantuml.com/plantuml",
  "record": "dev",
  "remoteAssets": false,
  "routerMode": "history",
  "selectable": false,
  "src": "sub/page1.md",
  "theme": "default",
  "themeConfig": {},
  "title": undefined,
  "titleTemplate": "%s - Slidev",
}
`;

exports[`md parser > multi-entries.md > features 1`] = `
{
  "katex": true,
  "mermaid": false,
  "monaco": false,
  "tweet": true,
}
`;

exports[`md parser > multi-entries.md > slides 1`] = `
[
  {
    "content": "
# Page 1
",
    "end": 2,
    "frontmatter": {
      "srcSequence": "sub/page1.md",
      "title": undefined,
    },
    "index": 0,
    "inline": {
      "content": "",
      "end": 4,
      "frontmatter": {
        "title": undefined,
      },
      "index": 0,
      "level": undefined,
      "note": undefined,
      "raw": "---
src: sub/page1.md
---
",
      "start": 0,
      "title": undefined,
    },
    "level": 1,
    "note": undefined,
    "raw": "---
srcSequence: sub/page1.md
---

# Page 1

",
    "source": {
      "content": "# Page 1",
      "end": 2,
      "frontmatter": {
        "title": "Page 1",
      },
      "index": 0,
      "level": 1,
      "note": undefined,
      "raw": "# Page 1
",
      "start": 0,
      "title": "Page 1",
    },
    "start": 0,
    "title": "Page 1",
  },
  {
    "content": "
# Page 2

<Tweet />
",
    "end": 8,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#2",
      "layout": "cover",
      "srcSequence": "sub/page2.md",
      "title": "Page 2",
    },
    "index": 1,
    "inline": {
      "content": "",
      "end": 9,
      "frontmatter": {
        "background": "https://sli.dev/demo-cover.png#2",
      },
      "index": 1,
      "level": undefined,
      "note": undefined,
      "raw": "---
src: sub/page2.md
background: https://sli.dev/demo-cover.png#2
---
",
      "start": 4,
      "title": undefined,
    },
    "level": 1,
    "note": undefined,
    "raw": "---
layout: cover
title: Page 2
background: https://sli.dev/demo-cover.png#2
srcSequence: sub/page2.md
---

# Page 2

<Tweet />

",
    "source": {
      "content": "# Page 2

<Tweet />",
      "end": 8,
      "frontmatter": {
        "layout": "cover",
        "title": "Page 2",
      },
      "index": 0,
      "level": 1,
      "note": undefined,
      "raw": "---
layout: cover
---

# Page 2

<Tweet />
",
      "start": 0,
      "title": "Page 2",
    },
    "start": 0,
    "title": "Page 2",
  },
  {
    "content": "
# Page 3
",
    "end": 2,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#34",
      "srcSequence": "sub/pages3-4.md",
      "title": "Page 3",
    },
    "index": 2,
    "inline": {
      "content": "",
      "end": 14,
      "frontmatter": {
        "background": "https://sli.dev/demo-cover.png#34",
      },
      "index": 2,
      "level": undefined,
      "note": undefined,
      "raw": "---
src: sub/pages3-4.md
background: https://sli.dev/demo-cover.png#34
---
",
      "start": 9,
      "title": undefined,
    },
    "level": 1,
    "note": undefined,
    "raw": "---
title: Page 3
background: https://sli.dev/demo-cover.png#34
srcSequence: sub/pages3-4.md
---

# Page 3

",
    "source": {
      "content": "# Page 3",
      "end": 2,
      "frontmatter": {
        "title": "Page 3",
      },
      "index": 0,
      "level": 1,
      "note": undefined,
      "raw": "# Page 3
",
      "start": 0,
      "title": "Page 3",
    },
    "start": 0,
    "title": "Page 3",
  },
  {
    "content": "
# Page 4

<Tweet />
",
    "end": 10,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#34",
      "layout": "cover",
      "srcSequence": "sub/pages3-4.md",
    },
    "index": 3,
    "level": 1,
    "note": undefined,
    "raw": "---
layout: cover
background: https://sli.dev/demo-cover.png#34
srcSequence: sub/pages3-4.md
---

# Page 4

<Tweet />

",
    "source": {
      "content": "# Page 4

<Tweet />",
      "end": 10,
      "frontmatter": {
        "layout": "cover",
      },
      "index": 1,
      "level": 1,
      "note": undefined,
      "raw": "---
layout: cover
---

# Page 4

<Tweet />
",
      "start": 2,
      "title": "Page 4",
    },
    "start": 2,
    "title": "Page 4",
  },
  {
    "content": "
# Page 1
",
    "end": 2,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#14",
      "srcSequence": "sub/nested1-4.md,sub/page1.md",
      "title": undefined,
    },
    "index": 4,
    "inline": {
      "content": "",
      "end": 19,
      "frontmatter": {
        "background": "https://sli.dev/demo-cover.png#14",
      },
      "index": 3,
      "level": undefined,
      "note": undefined,
      "raw": "---
src: sub/nested1-4.md
background: https://sli.dev/demo-cover.png#14
---
",
      "start": 14,
      "title": undefined,
    },
    "level": 1,
    "note": undefined,
    "raw": "---
background: https://sli.dev/demo-cover.png#14
srcSequence: sub/nested1-4.md,sub/page1.md
---

# Page 1

",
    "source": {
      "content": "# Page 1",
      "end": 2,
      "frontmatter": {
        "title": "Page 1",
      },
      "index": 0,
      "level": 1,
      "note": undefined,
      "raw": "# Page 1
",
      "start": 0,
      "title": "Page 1",
    },
    "start": 0,
    "title": "Page 1",
  },
  {
    "content": "
# Page 2

<Tweet />
",
    "end": 8,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#14",
      "layout": "cover",
      "srcSequence": "sub/nested1-4.md,sub/page2.md",
      "title": "Page 2",
    },
    "index": 5,
    "level": 1,
    "note": undefined,
    "raw": "---
layout: cover
title: Page 2
background: https://sli.dev/demo-cover.png#14
srcSequence: sub/nested1-4.md,sub/page2.md
---

# Page 2

<Tweet />

",
    "source": {
      "content": "# Page 2

<Tweet />",
      "end": 8,
      "frontmatter": {
        "layout": "cover",
        "title": "Page 2",
      },
      "index": 0,
      "level": 1,
      "note": undefined,
      "raw": "---
layout: cover
---

# Page 2

<Tweet />
",
      "start": 0,
      "title": "Page 2",
    },
    "start": 0,
    "title": "Page 2",
  },
  {
    "content": "
# Page 3
",
    "end": 2,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#14",
      "srcSequence": "sub/nested1-4.md,sub/pages3-4.md",
      "title": "Page 3",
    },
    "index": 6,
    "level": 1,
    "note": undefined,
    "raw": "---
title: Page 3
background: https://sli.dev/demo-cover.png#14
srcSequence: sub/nested1-4.md,sub/pages3-4.md
---

# Page 3

",
    "source": {
      "content": "# Page 3",
      "end": 2,
      "frontmatter": {
        "title": "Page 3",
      },
      "index": 0,
      "level": 1,
      "note": undefined,
      "raw": "# Page 3
",
      "start": 0,
      "title": "Page 3",
    },
    "start": 0,
    "title": "Page 3",
  },
  {
    "content": "
# Page 4

<Tweet />
",
    "end": 10,
    "frontmatter": {
      "background": "https://sli.dev/demo-cover.png#14",
      "layout": "cover",
      "srcSequence": "sub/nested1-4.md,sub/pages3-4.md",
    },
    "index": 7,
    "level": 1,
    "note": undefined,
    "raw": "---
layout: cover
background: https://sli.dev/demo-cover.png#14
srcSequence: sub/nested1-4.md,sub/pages3-4.md
---

# Page 4

<Tweet />

",
    "source": {
      "content": "# Page 4

<Tweet />",
      "end": 10,
      "frontmatter": {
        "layout": "cover",
      },
      "index": 1,
      "level": 1,
      "note": undefined,
      "raw": "---
layout: cover
---

# Page 4

<Tweet />
",
      "start": 2,
      "title": "Page 4",
    },
    "start": 2,
    "title": "Page 4",
  },
  {
    "content": "
# Inline Page

$x+2$
",
    "end": 25,
    "frontmatter": {},
    "index": 8,
    "level": 1,
    "note": undefined,
    "raw": "
# Inline Page

$x+2$

",
    "start": 20,
    "title": "Inline Page",
  },
]
`;
